### ISSUE TITLE
Provide a mechanism for limiting the number of expressions

### ISSUE THREAD

**Issue by lucperkins** [cel-rust/cel-rust#56]
Providing massive expressions is a good way to bring down a server. It's pretty trivial to, for example, write a little script that creates an "or" chain 100,000 Booleans long. Because of this, it'd be good to provide support for optionally limiting the number of expressions evaluated by a `Program`.

**Comment by lucperkins** [comment]
I have some ideas on how to make this happen and am happy to do the work. Just wanted to see if I encounter any resistance before I submit a PR 😄

**Comment by clarkmcc** [comment]
No resistance from me! That's actually something I've been thinking about recently. I know other CEL implementations do something similar to this, i.e. [cel-go](https://pkg.go.dev/github.com/google/cel-go/cel#CostLimit).

**Comment by lucperkins** [comment]
@clarkmcc Oh, nice! I was not aware of that implementation, but great to know. Glad you're amenable to this 😄 I'll submit a PR pretty soon here.

**Comment by clarkmcc** [comment]
Documenting for posterity: [this](https://github.com/google/cel-spec/blob/master/doc/langdef.md#performance) should probably be our guiding star for expression cost estimation/limiting.

**Comment by catkins** [comment]
Hi there 👋

Just wanted to +1 this issue. We have been looking at embedding this crate in Ruby on Rails applications with the [cel-rs-rb](https://github.com/buildkite/cel-rs-rb) Gem, and not having a way of limiting expressions or interrupt them would make this a lot more comfortable from a DOS point of view.

**Comment by alexsnaps** [comment]
I think that'd be covered if we'd implement the cost model feature, right?

**Comment by cryi** [comment]
Hi, I am also looking forward to seeing this implemented. I need both budgeting and timeouts.
We are migrating one of our internal projects from go to rust and while in general rust cel seems to be easier to work with I really lack ability to limit the execution.

**Comment by alexsnaps** [comment]
We probably want to start looking into opening the seam on the current API to be able to support this. 
iirc, the cost estimate runs on the `AST<Parsed>` as part or before the `AST<Checked>` pass in go?
Is that something we should mimic? Other features are available during that phase iirc, like optimizing passes on the `AST<Parsed>` - @howardjohn had started playing around with such concepts already too. 

Thinking out loud here, but deciding how we'd expose such features to a user would probably be good. I'd be tempted to keep the "AST states" as a concept (not necessarily as a type param like in "`AST<Parsed>`", could be, but haven't thought much about it). Thoughts?

**Comment by alexsnaps** [comment]
Also, on the time budget aspect, my first impression is that this is best kept "outside" the core. Not that we could support such a feature in this crate, but I think how that's implemented shouldn't impact the interpreter itself, should it? Thinking of e.g. [timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html)-like for async environment. 

Maybe have an interrupted flag _in the interpreter_, but would that even make sense... e.g. what about custom functions?

**Comment by alexsnaps** [comment]
Started [this discussion](https://github.com/cel-rust/cel-rust/discussions/323) on that wider subject

**Comment by cryi** [comment]
> Also, on the time budget aspect, my first impression is that this is best kept "outside" the core. Not that we could support such a feature in this crate, but I think how that's implemented shouldn't impact the interpreter itself, should it? Thinking of e.g. [timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html)-like for async environment.
> 
> Maybe have an interrupted flag _in the interpreter_, but would that even make sense... e.g. what about custom functions?

Yes cel-go interrupts the interpreter when the context expires. See [program.go#L459-L474]( https://github.com/cel-expr/cel-go/blob/6a93169e99dea3b6412af235711f61d7a47de835/cel/program.go#L459-L474). We need some kind of preemption to stop the interpreter. Timeout from ougside would mean leaving a timed out execution running in separate thread in the background.

**Comment by alexsnaps** [comment]
See #339

**Comment by catkins** [comment]
Nice work @alexsnaps. I had a look at the README update and that looks like exactly what I've been hoping for 🙇‍♂️

**Comment by alexsnaps** [comment]
@catkins good to know! I'll link this issue to that PR then. I'm hoping to then merge that as part of the next `v0.15+` release. 
Glad to hear, something like that is what you needed

**Comment by cryi** [comment]
Can't wait for 0.15.

**Comment by alexsnaps** [comment]
@cryi #342 is the big one slowing this all down, so if you want to review and/or play with the PR... You could make this move faster 😉

**Comment by cryi** [comment]
I see, @alexsnaps. To be honest, I got caught up with another project (don't want to spam links here, but feel free to check my commit history if you're curious). I need to fast-track it to 0.1.0 first, then I can get back to CEL and hopefully review this. No ETA right now, though.

**Comment by alexsnaps** [comment]
No worries @cryi - we all have our hands full, kinda good I guess!
